### ISSUE TITLE
Using Async Await?

### ISSUE THREAD

**Issue by Didza** [cosmicpython/code#61]
Is it possible to use async await, especially when accessing external resources like the db?

**Comment by hjwp** [comment]
i actually started investigating it in a branch somewhere once but i didn't finish it.

the tricky bit is figuring out how far down the stack you are going to allow async/await to encroach.  the pure domain code needs to be non-async, i would say.  but what about the Repository and the UnitOfWork?  Or the Service Layer?  i didn't decide on a final answer for that...

**Comment by Didza** [comment]
Theoretically it should be possible based on my understanding, it looks like the repository, service layer handlers to the entry points will be async and external calls awaited. Will investigate this further and document my findings here.

**Comment by nomhoi** [comment]
Please see my refactoring: https://github.com/nomhoi/cosmicpython-fastapi

**Comment by nomhoi** [comment]
There are many examples with asyncronous FastAPI for RealWorld project: 
https://github.com/search?q=realworld+fastapi

**Comment by Didza** [comment]
> Please see my refactoring: https://github.com/nomhoi/cosmicpython-fastapi

Thanks for making efforts in this direction will definitely be on the look out to see how you implement this include the message_bus as async. Thank you for sharing.

**Comment by nomhoi** [comment]
> Thanks for making efforts in this direction will definitely be on the look out to see how you implement this including the message_bus as async. Thank you for sharing.

I intend to complete all chapters.

**Comment by jalvespinto** [comment]
@nomhoi, tks for sharing this. While you are at it, it would be nice to consider issue [#23](https://github.com/cosmicpython/code/issues/23)

**Comment by nomhoi** [comment]
> @nomhoi, tks for sharing this. While you are at it, it would be nice to consider issue [#23](https://github.com/cosmicpython/code/issues/23)

Ok, a bit later.

**Comment by nomhoi** [comment]
All 13 chapters have been completed.

**Comment by nomhoi** [comment]
> @nomhoi, tks for sharing this. While you are at it, it would be nice to consider issue [#23](https://github.com/cosmicpython/code/issues/23)

Is the problem in this test?
https://github.com/xtaje/code/blob/2f260cba7eae3c6467c14befddfa92cc6d003c52/tests/e2e/test_queue_race.py

**Comment by jalvespinto** [comment]
I don't think so. If I understood it right the problem is when you have a server that don't spin up one thread per request. The reason been that the `bus` is a module variable and when in the same thread that would be a problem for the bus queue and the db session, since multiple requests would share the queue and session. In my case I am creating a new queue and uow on a per request basis, but I guess this is a problem for the tests on the project. I don't really know, because I am not running the projects code/tests.

**Comment by hjwp** [comment]
> Please see my refactoring: https://github.com/nomhoi/cosmicpython-fastapi

this is awesome!  well done :)

if i get a bit more time, i want to investigate some more patterns around this.  can we make it so that the repository is non-async somehow??  if you could somehow pre-load the state before entering the service layer, and have the outputs be some combination of state to modify (on a sqla session maybe?) plus events, then maybe the whole service layer could be sync somehow?  it wouldn't work for the UoW, but, idk, the dream is, the fewer unit tests use async, the better.

**Comment by nomhoi** [comment]
> 1. async tests are slower (they tend to start and stop an event loop, makes them an order of magnitude slower in my experience)

I used in conftest.py:

```python
@pytest.fixture(scope="session")
def event_loop():
    policy = asyncio.get_event_loop_policy()
    loop = policy.new_event_loop()
    yield loop
    loop.close()
```
Only one event loop will be created per a testing session:
https://pytest-asyncio.readthedocs.io/en/latest/reference/fixtures.html#event-loop
By default the scope is 'function'.

**Comment by nomhoi** [comment]
> if i get a bit more time, i want to investigate some more patterns around this. can we make it so that the repository is non-async somehow?? if you could somehow pre-load the state before entering the service layer, and have the outputs be some combination of state to modify (on a sqla session maybe?) plus events, then maybe the whole service layer could be sync somehow? it wouldn't work for the UoW, but, idk, the dream is, the fewer unit tests use async, the better.

I think we can get and pass running event loop to UoW and repository and run async session methods under it.

**Comment by nomhoi** [comment]
May be this will be usefull: https://returns.readthedocs.io/en/latest/pages/future.html

**Comment by nomhoi** [comment]
For example, in the branch sync_repository:
https://github.com/nomhoi/cosmicpython-fastapi/compare/main...sync_repository
I've change SqlAlchemyRepository and chapter_13_dependency_injection/tests/integration/test_repository.py.
SqlAlchemyRepository is sync now.
Please try:
```bash
make up
docker-compose run --rm --no-deps --entrypoint=pytest api /tests/integration/test_repository.py
```

**Comment by nomhoi** [comment]
This will work also:
```python
class SqlAlchemyRepository(AbstractRepository):
    def __init__(self, session: AsyncSession):
        super().__init__()
        self.session = session
        self.event_loop = asyncio.get_event_loop()
```

**Comment by nomhoi** [comment]
The same with UoW: https://github.com/nomhoi/cosmicpython-fastapi/commit/5e81bb9a7b26202ad2ea7f4da3b9041d73567b19

```bash
docker-compose run --rm --no-deps --entrypoint=pytest api /tests/integration/test_uow.py::test_uow_can_retrieve_a_batch_and_allocate_to_it
```

**Comment by dbaber** [comment]
Are derivatives/forks like hte fastapi one even legally allowed on github because the license is CC-By-ND?

**Comment by hjwp** [comment]
for the record, it's fine by me.  i'll see what i can do about the license...
